File: ros2_workspaces/ros2_from_scratch_by_Renard_ws/src/my_cpp_pkg/src/robot_hardware_status_publisher.cpp
#include "rclcpp/rclcpp.hpp"
// #include "example_interfaces/msg/int64.hpp" // interface for message
#include "my_robot_interfaces/msg/hardware_status.hpp" // interface for message

class RobotHardwareStatusPublisherNode : public rclcpp::Node  {

    public:
    // constructor
    RobotHardwareStatusPublisherNode() : Node("robot_hardware_status_node_cpp") {
        RCLCPP_INFO(this->get_logger(), "Creating %s...", this->get_name());
        hardware_status_publisher_ = this->create_publisher<my_robot_interfaces::msg::HardwareStatus>("robot_hardware_status", 10);
        hardware_status_timer_ = this->create_wall_timer(std::chrono::seconds(1), std::bind(&RobotHardwareStatusPublisherNode::publishNumber, this));
        RCLCPP_INFO(this->get_logger(), "%s created!!", this->get_name());
    }

    private:
    void publishNumber() {
        auto msg = my_robot_interfaces::msg::HardwareStatus();
        msg.temperature = 34.5;
        msg.version = 1;
        msg.are_motors_ready = true;
        msg.debug_message = "ready";
        hardware_status_publisher_->publish(msg);
    }
    int number_;
    /**
     * Publisher is a template class. 
     * It requires a message type to indicate the type of data it works with.
     * We use a shared pointer.
     */
    rclcpp::Publisher<my_robot_interfaces::msg::HardwareStatus>::SharedPtr hardware_status_publisher_;
    rclcpp::TimerBase::SharedPtr hardware_status_timer_;
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<RobotHardwareStatusPublisherNode>(); 
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: ros2_workspaces/Books/ros2_from_scratch_by_Renard/part2_ws/src/my_py_pkg/my_py_pkg/robot_hardware_status_publisher.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
# from example_interfaces.msg import Int64
from my_robot_interfaces.msg import HardwareStatus

"""
We have a publishers and a timer with a callback.
The timer creates an Int64 message and publishes it to the number topic.

This structure is very common in ROS2. Whenever we want to pubish data from a sensor, 
we use something similar.
"""
class RobotHardwareStatusPublisherNode(Node): 

    def __init__(self):
        super().__init__("robot_hardware_status_publisher_node") 
        # note super().init must run before we can use the logger
        self.get_logger().info("Starting "+self.get_name()+"...")
        self.number_ = 2
        # queue size is the number of messages that can be retained on the topic
        # if the publish rate exceeds the consumption rate
        self.number_publisher_ = self.create_publisher(HardwareStatus, "robot_hardware_status", 10)
        # Timer fires the callback each second
        self.number_timer_ = self.create_timer(1.0, self.publish_robot_hardware_status)
        self.get_logger().info(self.get_name()+" is started!!")

    def publish_robot_hardware_status(self):
        msg = HardwareStatus() # construct message instance
        msg.temperature = 34.5
        msg.version = 1
        msg.are_motors_ready = True
        msg.debug_message = "hardware is ready"
        self.number_publisher_.publish(msg)

# note that like c++, main belongs to the file, not a class.
def main(args=None):
    rclpy.init(args=args)
    node = RobotHardwareStatusPublisherNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()